gcc/rust/privacy/rust-visibility-resolver.cc

```diff
@@ -45,16 +45,56 @@ VisibilityResolver::go (HIR::Crate &crate)
     }
 }
 
-// FIXME: At this point in the pipeline, we should not be dealing with
-// `AST::SimplePath`s anymore! We need to be dealing with their "resolved
-// counterpart", so probably a NodeId/HirId/DefId.
+bool
+VisibilityResolver::resolve_module_path (const HIR::SimplePath &restriction,
+					 DefId &id)
+{
+  // We need, from the restriction, to figure out the actual Module it
+  // belongs to.
+
+  NodeId ast_node_id = restriction.get_mappings ().get_nodeid ();
+
+  auto invalid_path
+    = Error (restriction.get_locus (),
+	     "cannot use non-module path as privacy restrictor");
 
-// static bool
-// resolve_module_path (std::vector<HIR::Module> &module_stack,
-// 		     const AST::SimplePath &restriction, DefId &id)
-// {
-//   return false;
-// }
+  NodeId ref_node_id = UNKNOWN_NODEID;
+  if (!resolver.lookup_resolved_name (ast_node_id, &ref_node_id))
+    {
+      invalid_path.emit_error ();
+      return false;
+    }
+  // FIXME: Add a hint here if we can find the path in another scope, such as
+  // a type or something else
+  // TODO: For the hint, can we point to the original item's definition if
```

Comment:
What do you mean here?

Reply:
What I mean is that when we have the following code:
```rust
mod a {
     mod b {
         struct Inner;
    }
}
pub (in a::b::Inner) struct Outer;
````

Let's say we're in this pass, which does not care about whether or not the path used in the restriction is allowed (this is the role of the `PubRestrictedVisitor`. Here, we just want to resolve it to its proper `DefId`).

The visibility used here is a valid *path*: `a::b::Inner` is resolved properly in the SimplePath resolving. However, it does not point to a module, but to a type. The FIXME here says that we should think about hinting regarding the definition of `Inner`. Have a `rust_note` like "cannot use type as visibility restriction" and "item defined here". So we should look through the resolved type scope as well, or check if the visibility used actually refers to a function and not a module, etc etc.

I think I quite know how to do it, it's mostly about doing it piece by piece and focusing on semantics rather than user experience regarding visibilities for now :)

Reply:
Ah ok i see what you mean. That should be ok if you fail to lookup in the name scope you can lookup in the type scope, do we care about the label scope? Then the mappings class we could add more enum's so we could switch on the types or add a visitor or something. I did something similar this over here:

https://github.com/Rust-GCC/gccrs/blob/9ea940e4dcabbf99fbb44c125a0af7cf82e48146/gcc/rust/typecheck/rust-hir-inherent-impl-overlap.h#L104